ios/NativeViews/IpfsImageView.m

```diff
@@ -0,0 +1,460 @@
+  for (IpfsImageSource *source in _imageSources) {
+    CGSize imgSize = source.size;
+    const CGFloat imagePixels =
+    imgSize.width * imgSize.height * source.scale * source.scale;
+    const CGFloat fit = ABS(1 - (imagePixels / targetImagePixels));
+
+    if (fit < bestFit) {
+      bestFit = fit;
+      bestSource = source;
+    }
+  }
+  return bestSource;
+}
+
+- (BOOL)shouldReloadImageSourceAfterResize
+{
+  // If capInsets are set, image doesn't need reloading when resized
+  return UIEdgeInsetsEqualToEdgeInsets(_capInsets, UIEdgeInsetsZero);
+}
+
+- (BOOL)shouldChangeImageSource
+{
+  // We need to reload if the desired image source is different from the current image
+  // source AND the image load that's pending
+  IpfsImageSource *desiredImageSource = [self imageSourceForSize:self.frame.size];
+  return ![desiredImageSource isEqual:_imageSource] &&
+  ![desiredImageSource isEqual:_pendingImageSource];
+}
+
+- (void)reloadImage
+{
+  [self cancelImageLoad];
+  _needsReload = NO;
+
+  IpfsImageSource *source = [self imageSourceForSize:self.frame.size];
+  _pendingImageSource = source;
+
+  if (source && self.frame.size.width > 0 && self.frame.size.height > 0) {
+    if (_onLoadStart) {
+      _onLoadStart(nil);
+    }
+
+    RCTImageLoaderProgressBlock progressHandler = nil;
+    if (_onProgress) {
+      progressHandler = ^(int64_t loaded, int64_t total) {
+        self->_onProgress(@{
+                            @"loaded": @((double)loaded),
+                            @"total": @((double)total),
+                            });
+      };
+    }
+
+    __weak IpfsImageView *weakSelf = self;
+    RCTImageLoaderPartialLoadBlock partialLoadHandler = ^(UIImage *image) {
+      [weakSelf imageLoaderLoadedImage:image error:nil forImageSource:source partial:YES];
+    };
+
+    CGSize imageSize = self.bounds.size;
+    CGFloat imageScale = RCTScreenScale();
+    if (!UIEdgeInsetsEqualToEdgeInsets(_capInsets, UIEdgeInsetsZero)) {
+      // Don't resize images that use capInsets
+      imageSize = CGSizeZero;
+      imageScale = source.scale;
+    }
+
+    RCTImageLoaderCompletionBlock completionHandler = ^(NSError *error, UIImage *loadedImage) {
+      [weakSelf imageLoaderLoadedImage:loadedImage error:error forImageSource:source partial:NO];
+    };
+
+//    _reloadImageCancellationBlock =
+//    [_bridge.imageLoader loadImageWithURLRequest:source.request
+//                                            size:imageSize
+//                                           scale:imageScale
+//                                         clipped:NO
+//                                      resizeMode:_resizeMode
+//                                   progressBlock:progressHandler
+//                                partialLoadBlock:partialLoadHandler
+//                                 completionBlock:completionHandler];
+
+    NSError *error;
+    UIImage *image;
+    NSString *base64String = [_bridge.ipfs _getHashData:source.hashPath error:&error];
```

Comment:
Even though our API for loading image data is synchronous, I left all the async handling from the `URLRequest`-based `Image` code just in case we switch our API to async some day. Seems like that might be a good idea.